[14.0] helpdesk_mgmt: Forward port multiple 12.0 commits

ypapouin opened the pull request:
See commits for details.
This is a pack multiple commits authored by @eLBati, fixed and rewritten to be compatible with Odoo 14.0.

- Set "e-mail" channel when ticket generated from fetchmail cron task.
- Add a list of attachments in the ticket portal view.
- When a ticket is created by a customer with attachments, an access token is automatically generated.
- Allow a portal user to close its ticket if the final stage is set with `portal_user_can_close`.
- Sequence field added to allow kanban reordering.
- E-mail templates are simplified, they now all use the `mail.mail_notification_light` layout to have a consistent design.
- Send a notification to the partner each time the ticket stage changes.

eLBati committed:
[IMP] helpdesk_mgmt: Set e-mail channel when creating ticket by fetchmail

Forward port of fcf0a41bbf4099d7926cf542402d23970a3dcf39

eLBati committed:
[IMP] helpdesk_mgmt: Show attachments on portal's ticket view

Forward port of 656c2d933586fcb08a50488ee5b0cd27b56dfe03

Note that the commit fc3905eaa2330e7d70714f9a614eb2b48d4d173f (allow portal users to write tickets) has not been forward ported as it is not necessary

eLBati committed:
[IMP] helpdesk_mgmt: Adding `portal_user_can_close` to show closing buttons on portal

Forward port of 6eb1124dfe631ef16cf8669fefb7d51479a3d0d0 with minor fixes

eLBati committed:
[IMP] helpdesk_mgmt: e-mail templates
- Better style
- Avoid email_cc (error "message without recipients")
- Use coherent partner/user reference
- Access button from e-mail notifications

Forward port of 98e5ef7f4a2e074283da3674654f662a28f02eed

Please note that the commit a5041c25a79cd9620d30cc09dd9741f46aef7e76 (An e-mail is sent to the ticket after each create/edit) has not been forward ported because the assignment notify is done internally by odoo in `_message_auto_subscribe_followers` from `mail.thread`

eLBati committed:
[IMP] helpdesk_mgmt: Sequence field added to sort tickets in kanban view, also considering priority

Forward port of fe7a9eff891b50c8c040df1ead5f4bfde0d8843f

eLBati committed:
[IMP] helpdesk_mgmt: Test access added

Forward port of dfbc3698e76b3a1330f339ca78b49557b78c05ed

ypapouin committed:
[IMP] helpdesk_mgmt: Use email layout to simplify all templates

ypapouin committed:
[IMP] helpdesk_mgmt: Generate `access_token` for all attachments added on ticket submit

Otherwise they stay "private" and are not displayed on the portal.

ypapouin committed:
[IMP] helpdesk_mgmt: Send a notification to the partner when the ticket stage has changed

lk-eska commented:
when do we expect to merge these changes?

HekkiMelody commented on helpdesk_mgmt/controllers/main.py:
```diff
@@ -90,15 +98,18 @@ def submit_ticket(self, **kw):
         new_ticket = request.env["helpdesk.ticket"].sudo().create(vals)
         new_ticket.message_subscribe(partner_ids=request.env.user.partner_id.ids)
         if kw.get("attachment"):
+            IrAttachment = request.env["ir.attachment"]
+            attachment_ids = IrAttachment
```
what's the point of this?
```suggestion
            attachment_ids = request.env["ir.attachment"]
```

ypapouin replied:
`IrAttachment` is used in the create statement below, that's why the model is set once outside the loop

HekkiMelody reviewed (approved):
Code review

HekkiMelody commented on helpdesk_mgmt/models/helpdesk_ticket.py:
```diff
@@ -103,8 +105,23 @@ def _read_group_stage_ids(self, stages, domain, order):
 
+    def send_partner_mail(self):
+        created_ticket_template = self.env.ref(
+            "helpdesk_mgmt.created_ticket_template", raise_if_not_found=False
```
This template is missing

HekkiMelody reviewed (changes requested):
Really want these changes merged.

There are some issues. For example the customer can add some attachments and see them, but there doesn't seem to be a way for the internal operator to upload attachments from the user view and show them for the customer in the portal view. 

Also, I think the "created" email template should not be forcefully sent in the create method, but should be attached to the "new" state and work like every other email template.

That said, I'd rather this be merged and these issues fixes in future PRs, personally. The only thing that's blocking IMO is the missing email template for new tickets.

ypapouin commented:
@aleuffre, the missing template is re-added.

About this, I tried to use the built-in way to send email notification on ticket creation by assigning `helpdesk_ticket_stage_new.mail_template_id` but it's not working when the ticket is created from the portal or duplicated from the backend, I've added a comment in the `send_partner_mail` function (It could be related to the odoo framework).

About attachments, the chatter allows the internal users to add public attachments from the backend (using `Send message`) and since Odoo 14.0, any users (portal/internal) can add public attachments from the portal.
The attachment list is still there as a feedback.

Also, please give me some time, I will try to increase code coverage before merging.

ypapouin committed:
[IMP] helpdesk_mgmt: Test closing ticket

- `portal_user_can_close` is now set to `done` and `cancelled` stages by default
- `/ticket/close` has been replaced with `/close/ticket` to be consistent with other routes

ypapouin committed:
[IMP] helpdesk_mgmt: Test submit with attachments

ypapouin committed:
[IMP] helpdesk_mgmt: Test ticket created after a mail fetch

pedrobaeza commented:
Too much things in one PR. This should be split by each function to be discussed and reviewed isolatedly, or there are few chances of this being merged as is.

ypapouin commented:
@pedrobaeza, I understand your position but most of these commits are already merged in the 12.0 branch, its like a migration update, others one are tests.

pedrobaeza commented:
I use the module since 13.0, so I didn't take part on the decisions made on v12, so I'm still reluctant to approve this on batch.

HekkiMelody commented:
I also need these commits on v14. Will start opening some PRs separately if the whole thing won't be accepted.

ypapouin commented:
@pedrobaeza as your request, this PR has been split into multiple PRs.
So I'm closing this one.
